### ISSUE TITLE
Cloud syncing

### ISSUE THREAD

**Issue by rileymacisaac** [timkurvers/valheim-macos#43]
Hey there,

I'm loving this, and it runs great. I even have mods working through Bepinex.

But I can't figure out how to get cloud syncing working at all. I managed to get my character loaded and working, but every time I close the game steam throws an error saying there is a conflict with the cloud sync.

Any advice or help would be awesome.

**Comment by AudioCow** [comment]
The problem is that Steam Cloud tries to write the cloud save file to the local Windows AppData folder.  Since the directory is not found, it causes issues.

**Comment by LunaticZorr** [comment]
I'm facing the same issue. I'd love to get Cloud Sync working because I play on my Mac and sometimes move to a different PC running windows. The logs file shows the following:

```
[2024-01-03 18:12:38] [AppID 892970] YldWriteCacheDirectoryToFile - saved to '/Users/kevmas/Library/Application Support/Steam/userdata/95862329/892970/remotecache.vdf'
[2024-01-03 18:13:02] [AppID 892970] YldWriteCacheDirectoryToFile - saved to '/Users/kevmas/Library/Application Support/Steam/userdata/95862329/892970/remotecache.vdf'
[2024-01-03 18:13:22] [AppID 892970] YldWriteCacheDirectoryToFile - saved to '/Users/kevmas/Library/Application Support/Steam/userdata/95862329/892970/remotecache.vdf'
[2024-01-03 18:13:41] [AppID 892970] Starting sync (AC Exit,Sync Disabled,)
[2024-01-03 18:13:41] [AppID 892970] Running AutoCloud on exit. Looking for new and updated files
[2024-01-03 18:13:41] [AppID 892970]     Skipping rule 0 because ResolvePath (WinAppDataLocalLow, IronGate/Valheim/characters) failed
[2024-01-03 18:13:41] [AppID 892970]     Skipping rule 1 because ResolvePath (WinAppDataLocalLow, IronGate/Valheim/worlds) failed
[2024-01-03 18:13:41] [AppID 892970]     No launch record found. Skipping looking for files while the game was running
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/characters/lunatic zorr.fch.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/characters/lunatic zorr_backup_20220623-190445.fch) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/Attempt2.db.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/Attempt2.fwl.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/Attempt2_backup_20220623-190446.db) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/Attempt2_backup_20220623-190446.fwl) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/DaBomb.db) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/DaBomb.fwl) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/DaBomb.fwl.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/eeeweweweew.db) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/eeeweweweew.db.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/eeeweweweew.fwl) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/eeeweweweew.fwl.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaticZorrTestMola.db) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaticZorrTestMola.db.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaticZorrTestMola.fwl) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaticZorrTestMola.fwl.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaZorrTest2.db) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaZorrTest2.db.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaZorrTest2.fwl) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] FileForget(IronGate/Valheim/worlds/LunaZorrTest2.fwl.old) - file has pending changes in cloud; ignoring forget so as not to create a conflict.
[2024-01-03 18:13:41] [AppID 892970] AutoCloud complete
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Unable to resolve path with root WinAppDataLocalLow (12)
[2024-01-03 18:13:41] [AppID 892970] Failed sync for 'AC Exit,Sync Disabled,' [login=false][offlineMode=false]
[2024-01-03 18:13:41] [AppID 892970] YldWriteCacheDirectoryToFile - saved to '/Users/kevmas/Library/Application Support/Steam/userdata/95862329/892970/remotecache.vdf'
```

**Comment by LunaticZorr** [comment]
Ok another quick update. I've managed to get it working.

1. I closed the game and Steam.
2. I renamed ` ~/Library/Application Support/Steam/userdata/<user-id>/892970` to ` ~/Library/Application Support/Steam/userdata/<user-id>/892970.back`
3. I reopened steam and Valheim.
4. I created a new character and a new world. This re-created the ` ~/Library/Application Support/Steam/userdata/<user-id>/892970/characters` and ` ~/Library/Application Support/Steam/userdata/<user-id>/892970/worlds` folders.
5. I closed the game and checked the logs.

This triggered the new world and characters to be uploaded, the old ones to be deleted (although they weren't there anymore ?????)

```
[2024-01-04 12:37:06] [AppID 892970] YldWriteCacheDirectoryToFile - saved to '/Users/kevmas/Library/Application Support/Steam/userdata/95862329/892970/remotecache.vdf'
[2024-01-04 12:37:06] [AppID 892970] YldWriteCacheDirectoryToFile - saved to '/Users/kevmas/Library/Application Support/Steam/userdata/95862329/892970/remotecache.vdf'
[2024-01-04 12:38:02] [AppID 892970] Starting sync (up,AC Exit,)
[2024-01-04 12:38:02] [AppID 892970] Running AutoCloud on exit. Looking for new and updated files
[2024-01-04 12:38:02] [AppID 892970]     Skipping rule 0 because ResolvePath (WinAppDataLocalLow, IronGate/Valheim/characters) failed
[2024-01-04 12:38:02] [AppID 892970]     Skipping rule 1 because ResolvePath (WinAppDataLocalLow, IronGate/Valheim/worlds) failed
[2024-01-04 12:38:02] [AppID 892970]     No launch record found. Skipping looking for files while the game was running
[2024-01-04 12:38:02] [AppID 892970] AutoCloud complete
[2024-01-04 12:38:03] [AppID 892970] Attempting to sync from local change number '0' to global change number '9415738457201706157'
[2024-01-04 12:38:03] [AppID 892970] Need to delete file characters/qwerty.fch.new
[2024-01-04 12:38:03] [AppID 892970] Need to delete file characters/qwerty.fch
[2024-01-04 12:38:03] [AppID 892970] Need to upload file characters/qwerty_backup_auto-20240104123510.fch
[2024-01-04 12:38:03] [AppID 892970] Need to delete file characters/test-fourth-jan.fch.new
[2024-01-04 12:38:03] [AppID 892970] Need to upload file characters/test-fourth-jan.fch
[2024-01-04 12:38:03] [AppID 892970] Need to upload file characters/test-fourth-jan_backup_auto-20240104123541.fch
[2024-01-04 12:38:03] [AppID 892970] Need to upload file worlds/test4thjan.fwl
[2024-01-04 12:38:03] [AppID 892970] Need to upload file worlds/test4thjan_backup_auto-20240104123706.fwl

[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'worlds/Attempt2.fwl.old' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'serverlist/recent' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:03] [AppID 892970] Download OK for file worlds/Attempt2.fwl.old
[2024-01-04 12:38:03] [AppID 892970] Download OK for file serverlist/recent
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'serverlist/favorite' from host 'steamcloudams2.blob.core.windows.net' - Success.
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'worlds/Attempt2_backup_20231206-003046.fwl' from host 'steamcloudlondon.blob.core.windows.net' - Success.
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'worlds/Attempt2.fwl' from host 'steamcloudams2.blob.core.windows.net' - Success.
[2024-01-04 12:38:03] [AppID 892970] Download OK for file worlds/Attempt2_backup_20231206-003046.fwl
[2024-01-04 12:38:03] [AppID 892970] Download OK for file serverlist/favorite
[2024-01-04 12:38:03] [AppID 892970] Download OK for file worlds/Attempt2.fwl
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'serverlist/favorite.old' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'serverlist/recent.old' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:03] [AppID 892970] Download OK for file serverlist/favorite.old
[2024-01-04 12:38:03] [AppID 892970] Download OK for file serverlist/recent.old
[2024-01-04 12:38:03] [AppID 892970] HTTP download for file 'worlds/qwrhfgtfdgjftgt.fwl' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:03] [AppID 892970] Download OK for file worlds/qwrhfgtfdgjftgt.fwl

[2024-01-04 12:38:04] [AppID 892970] HTTP download for file 'characters/lunatic zorr.fch.old' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:04] [AppID 892970] HTTP download for file 'worlds/qwrhfgtfdgjftgt.db' from host 'steamcloudlondon.blob.core.windows.net' - Success.
[2024-01-04 12:38:04] [AppID 892970] Download OK for file characters/lunatic zorr.fch.old
[2024-01-04 12:38:04] [AppID 892970] Download OK for file worlds/qwrhfgtfdgjftgt.db
[2024-01-04 12:38:04] [AppID 892970] HTTP download for file 'worlds/Attempt2_backup_20231206-003046.db' from host 'steamcloud-london.storage.googleapis.com' - Success.
[2024-01-04 12:38:05] [AppID 892970] Download OK for file worlds/Attempt2_backup_20231206-003046.db
[2024-01-04 12:38:05] [AppID 892970] HTTP download for file 'worlds/Attempt2_backup_auto-20231206233902.db' from host 'steamcloudlondon.blob.core.windows.net' - Success.
[2024-01-04 12:38:05] [AppID 892970] HTTP download for file 'worlds/Attempt2.db' from host 'steamcloud-eu-ams.storage.googleapis.com' - Success.
[2024-01-04 12:38:05] [AppID 892970] Download OK for file worlds/Attempt2_backup_auto-20231206233902.db
[2024-01-04 12:38:05] [AppID 892970] Download OK for file worlds/Attempt2.db
[2024-01-04 12:38:05] [AppID 892970] HTTP download for file 'worlds/Attempt2.db.old' from host 'steamcloudlondon.blob.core.windows.net' - Success.
[2024-01-04 12:38:05] [AppID 892970] Download OK for file worlds/Attempt2.db.old
[2024-01-04 12:38:05] [AppID 892970] Download complete, result OK
[2024-01-04 12:38:05] [AppID 892970] Successfully synced to ChangeNumber 9415738457201706157
[2024-01-04 12:38:05] [AppID 892970] Upload batch initiated, ID: 5930745085897927508, ChangeNumber: 14957474717695427741, App Build ID: 0
[2024-01-04 12:38:06] [AppID 892970] Delete OK for file characters/qwerty.fch.new
[2024-01-04 12:38:06] [AppID 892970] Delete OK for file characters/qwerty.fch
[2024-01-04 12:38:06] [AppID 892970] Delete OK for file characters/test-fourth-jan.fch.new
[2024-01-04 12:38:06] [AppID 892970] HTTP upload for file 'worlds/test4thjan.fwl' beginning . . .
[2024-01-04 12:38:06] [AppID 892970] HTTP upload for file 'worlds/test4thjan_backup_auto-20240104123706.fwl' beginning . . .
[2024-01-04 12:38:06] [AppID 892970] HTTP upload for file 'characters/qwerty_backup_auto-20240104123510.fch' beginning . . .
[2024-01-04 12:38:06] [AppID 892970] HTTP upload for file 'characters/test-fourth-jan.fch' beginning . . .
[2024-01-04 12:38:06] [AppID 892970] HTTP upload for file 'characters/test-fourth-jan_backup_auto-20240104123541.fch' beginning

[2024-01-04 12:38:06] [AppID 892970] Upload OK for file characters/qwerty_backup_auto-20240104123510.fch
[2024-01-04 12:38:06] [AppID 892970] Upload OK for file characters/test-fourth-jan.fch
[2024-01-04 12:38:06] [AppID 892970] Upload OK for file characters/test-fourth-jan_backup_auto-20240104123541.fch
[2024-01-04 12:38:07] [AppID 892970] Upload OK for file worlds/test4thjan_backup_auto-20240104123706.fwl
[2024-01-04 12:38:07] [AppID 892970] Upload complete, result OK
```

I then 

1. Copied the characters I wanted from the `.back` folder to the new folder.
2. Started Steam and Valheim.
3. Checked the saves in game and for my newly copied characters clicked `Restore` in game.
4. Made sure the saves are set to Cloud Save.
5. Closed the game.

This triggered the upload for these characters:

```
[2024-01-04 12:45:39] [AppID 892970] Need to upload file characters/lunatic zorr_backup_20240103-173550.fch
[2024-01-04 12:45:39] [AppID 892970] Need to upload file characters/shinywhitey_backup_restore-20240104-124402.fch
[2024-01-04 12:45:39] [AppID 892970] Need to upload file characters/shinywhitey_backup_20240104-124413.fch

[2024-01-04 12:45:39] [AppID 892970] Need to upload file characters/lunatic zorr.fch
[2024-01-04 12:45:39] [AppID 892970] Need to upload file characters/shinywhitey.fch
[2024-01-04 12:45:39] [AppID 892970] Need to upload file characters/shinywhitey.fch.old
[2024-01-04 12:45:39] [AppID 892970] Upload batch initiated, ID: 3182412891120532865, ChangeNumber: 11246540022575591647, App Build ID: 0
[2024-01-04 12:45:39] [AppID 892970] HTTP upload for file 'characters/shinywhitey_backup_restore-20240104-124402.fch' beginning . . .
[2024-01-04 12:45:39] [AppID 892970] HTTP upload for file 'characters/shinywhitey_backup_20240104-124413.fch' beginning . . .
[2024-01-04 12:45:39] [AppID 892970] HTTP upload for file 'characters/shinywhitey.fch' beginning . . .
[2024-01-04 12:45:39] [AppID 892970] HTTP upload for file 'characters/shinywhitey.fch.old' beginning . . .
[2024-01-04 12:45:39] [AppID 892970] HTTP upload for file 'characters/lunatic zorr_backup_20240103-173550.fch' beginning . . .

[2024-01-04 12:45:40] [AppID 892970] Upload OK for file characters/shinywhitey_backup_restore-20240104-124402.fch
[2024-01-04 12:45:40] [AppID 892970] Upload OK for file characters/shinywhitey.fch
[2024-01-04 12:45:40] [AppID 892970] Upload OK for file characters/shinywhitey_backup_20240104-124413.fch
[2024-01-04 12:45:40] [AppID 892970] Upload OK for file characters/lunatic zorr_backup_20240103-173550.fch
[2024-01-04 12:45:40] [AppID 892970] Upload OK for file characters/shinywhitey.fch.old

[2024-01-04 12:45:41] [AppID 892970] Upload OK for file characters/lunatic zorr.fch
[2024-01-04 12:45:42] [AppID 892970] Upload complete, result OK
```

I checked my cloud saves in the Steam website, and confirmed that the saves where there.
I opened Valheim on the Windows PC and checked the saves. All there.
I could load the characters :)

**Comment by rileymacisaac** [comment]
@LunaticZorr This absolutely worked, great job! In hindsight I probably should have tried this haha, oh well.

Thanks for that!

Btw – not sure if it'll mess something else up later but I didn't have to copy the characters or worlds back over afterwards. They were all there and up to date.

**Comment by timkurvers** [comment]
Closing this issue as the official macOS client is now available! 🥳 (see #104)
